python/amici/ode_export.py

```diff
@@ -897,19 +905,43 @@ def dx_dt(x_index, x_Sw):
                     si.species_compartment[x_index])]
                 return x_Sw/v
 
+        # create dynmics without respecting conservation laws first
         Sw = (MutableDenseMatrix(si.stoichiometric_matrix)
-              * MutableDenseMatrix(self.sym('w')))
+              * MutableDenseMatrix(fluxes))
         symbols['species']['dt'] = sp.Matrix([Sw.row(x_index).applyfunc(
             lambda x_Sw: dx_dt(x_index, x_Sw))
             for x_index in range(Sw.rows)])
 
+        # create all basic components of the ODE model and add them.
         for symbol in [s for s in symbols if s != 'my']:
             # transform dict of lists into a list of dicts
             protos = [dict(zip(symbols[symbol], t))
                       for t in zip(*symbols[symbol].values())]
             for proto in protos:
                 self.add_component(symbol_to_type[symbol](**proto))
 
+        # process conservation laws
+        if compute_cls:
+            si.process_conservation_laws(self)
+
+        # set derivatives of xdot, this circumvents regular computation. we
+        # do this as we can save a substantial amount of computations by
+        # knowing the right solutions here
+        nx_solver = si.stoichiometric_matrix.shape[0]
+        self._eqs['dxdotdx'] = sp.zeros(nx_solver)
+
+        nw = len(self._expressions)
+        # append zero rows for conservation law `w`s, note that
+        # _process_conservation_laws is called after the fluxes are added as
+        # expressions, if this ordering needs to be changed, this will have
+        # to be adapted.
+        self._eqs['dxdotdw'] = si.stoichiometric_matrix.row_join(
+            sp.zeros(nx_solver, nw-nr)
+        )
```

Comment:
We can either shorten the expression or increase dxdotdw.
I suggested to do the former. What's the advantage of doing the latter?

Reply:
It doesn't make sense to add the conservation laws as expressions and then remove them from the expressions again. For this very specific setup of the sbml_import it currently doesn't matter (because those species are not used outside the flux vector, where they are currently flattened out).

Adding stuff and removing it again is generally very bad practice as it increases the total amount of code and number of functions/files you have to look at in order to figure out what is actually happening. If you don't want to add them to the expressions, implement that in add_conservation_laws.

Reply:
This is why I was considering to remove the adding of CLs to the expressions. I just didn't do this because I was sure it was there for good purpose and probably needed for pySB. The SBML-Importer doesn't need it, and I also don't see, how it would need it, if real CLs were implemented...

From that perspective, I was left with: either (1) removing them right now directly or (2) adding zeros to dxdotdw and remove them by multiplying them with zeros.
(2) doesn't seem more elegant/any better practice to me than (1), as it also adds stuff just to remove (indirectly)later and moreover increases the size of variables used in C++(such as w) by entries, which are never used and hence adds (even if only minimal) computational overhead.

Reply:
In other words, what is done with `self._eqs['dxdotdx']` and `self._eqs['dxdotdw']` is absolutely not how things are supposed to be done. But it's one way to do it as efficiently as it is currently done. 

If there already is a function that does something similar to what you want, change that function and don't try to wiggle your away around outside of it.

Reply:
> This is why I was considering to remove the adding of CLs to the expressions. I just didn't do this because I was sure it was there for good purpose and probably needed for pySB. The SBML-Importer doesn't need it, and I also don't see, how it would need it, if real CLs were implemented...
> 
> From that perspective, I was left with: either (1) removing them right now directly or (2) adding zeros to dxdotdw and remove them by multiplying them with zeros.
> (2) doesn't seem more elegant/any better practice to me than (1), as it also adds stuff just to remove (indirectly)later and moreover increases the size of variables used in C++(such as w) by entries, which are never used and hence adds (even if only minimal) computational overhead.

The fact that all fluxes are captured in `w` and that state variables **only** occur in `w` is absolutely unique to the python-sbml importer (and bound not to be true at some point as we add features to the sbml import). For matlab models as well as pysb models states can and do occur outside of `w`, which means that having them defined as expression is necessary (although conservation laws don't exist in matlab models).

Its fine to change this implementation, but you have to do it in the right place. So where are those Expressions added (this is the place you will want to edit)? `ODE_model.add_conservation_law()`. Turns out that is a general function that is implemented independent of SBML/PySB import. If you change something there, this also changes stuff for the pysb import. This is intended like that. This function is not taylored to the pysb import, but to what we can generally assume about models defined in ODE_model. Changing that function will mess up a lot of stuff you don't want to deal with, so its way easier to just add the couple of entries to `w` and live with the minimal overhead, having general purpose implementation means that we have to do that from time to time.